### PR TITLE
[System.Numerics] Cross Product for Vector2 and Vector4

### LINKED ISSUE #28731: (System.Numerics) Cross Product for Vector2 and Vector4
I am proposing that a Cross Product function be added to both `Vector2` and `Vector4` of System.Numerics because its use in geometry is so common.

### Rationale and Usage
The structs, `Vector2`, `Vector3`, and `Vector4` are intended to model geometry functions that often need to be performed quickly and which would greatly benefit from SIMD inlining. Dot Product is just as popular and it is properly implemented in these structs. Cross Product is only possible for vectors of length 3 (yes, and 7, but that is not important right now). However, Vector2 and Vector4 are essentially modelling vectors in 3D space. In `Vector2` the cross product is commonly used for functions like offsetting, for determining if the next edge in a polygon turns right or left (or convex or concave), or determining overlap/collisions. A cross-product of two `Vector2` 's would always result in an out-of-plane or z-direction, but it's value and sign is very useful and just as quick to process as Dot.
`Vector4` is defined for use as a homogeneous 3D coordinate vector, and so it will be necessary for a `Cross` function to exist for it just as much as for `Vector3`.

### Proposed Changes And API
The addition of two new methods are quite simple and the functions that these invoke are nothing other than add (`+`) and multiply (`*`).
To Vector2_Instrinsics.cs (under Dot() line 109):

```cs
         /// <summary>
        /// Returns the z-value of the cross product of two vectors.
        /// Since the Vector2 is in the x-y plane, a 3D cross product
        /// only produces the z-value
        /// </summary>
        /// <param name="value1">The first vector.</param>
        /// <param name="value2">The second vector.</param>
        /// <returns>The value of the z-coordinate from the cross product.</returns>
        [JitIntrinsic]
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static float Cross(Vector2 value1, Vector2 value2)
        {
            return value1.X * value2.Y 
                   - value1.Y * value2.X;
        }
```

to Vector4_Instrinsics.cs (under Dot() line 157):

```cs
        /// <summary>
        /// Computes the cross product of two vectors. For homogeneous
        /// coordinates, the product of the weights is the new weight
        /// for the resulting product.
        /// </summary>
        /// <param name="vector1">The first vector.</param>
        /// <param name="vector2">The second vector.</param>
        /// <returns>The cross product.</returns>
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static Vector4 Cross(Vector4 vector1, Vector4 vector2)
        {
            return new Vector4(
                vector1.Y * vector2.Z - vector1.Z * vector2.Y,
                vector1.Z * vector2.X - vector1.X * vector2.Z,
                vector1.X * vector2.Y - vector1.Y * vector2.X,
                vector1.W * vector2.W);
        }
```

### Other Details
* One will note that the 'Cross' for 'Vector2' looks like the 'Dot' except the values have moved around. Yes, essentially the 'Cross' for 'Vector2' is the same as taking the 'Dot' with a vector that has been rotate 90-degrees from the original. As such, this is the work around to invoking geometric functions - create a new vector 90-degrees from the original, and then use `Dot`. This process is not intuitive and requires extra steps. A simple function, like the one shown above, is both more intuitive and quicker.
* The proposed `Cross` function for `Vector4` is nearly the same as that for `Vector3` with the addition of the fourth value which is the product of the original two w's. This can be derived by symbolically performing the cross product for `Vector3` with values `[x_1/w_1, y_1/w_1, z_1/w_1]` and `[x_2/w_2, y_2/w_2, z_2/w_2]`.

### Open Questions
* Will the name `Cross` be misleading for `Vector2`? If so, one could name it `CrossZValue`, but I think this is an unwarranted complication. In implementing 2D algorithms, the literature has called this the cross product and one would write the code to mimic the presentation of scientific papers. However, since the result is a single scalar (float), it could be confusing as to why a `Cross` product function is not returning a new vector.

### Pull Request
A PR with the proposed changes is forthcoming.

### Updates

### PR DESCRIPTION
**Pull request by AlexRadch**
Close https://github.com/dotnet/runtime/issues/28731 Issue.

### CODE REVIEW

**Comment by EgorBo** [comment]
Can you post codegen examples for these functions you added? Presumably, the API proposal implies JIT support for these which is not done as part of this PR

**Comment by AlexRadch** [comment]
> Can you post codegen examples for these functions you added? Presumably, the API proposal implies JIT support for these which is not done as part of this PR

Is there any instruction on how to get generated code?

**Comment by EgorBo** [comment]
I assume we can handle this separately in JIT once this merged

**Comment by huoyaoyuan** [comment]
Are there any corner cases we need to consider, like +0/-0? I assume we should avoid this to achieve best performance, but hardware intrinsic may provide different result.

**Comment by AlexRadch** [comment]
> Are there any corner cases we need to consider, like +0/-0? I assume we should avoid this to achieve best performance, but hardware intrinsic may provide different result.

`Vector2.Cross` have the same code as `Vector2.Dot` but with a rotated second vector.
`Vector4.Cross` has a similar code as `Vector3.Cross` and W components multiplication.

I think we do not need to consider any corner cases there.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs [hunk 1]
```diff
@@ -326,6 +326,26 @@ public static float Dot(Vector2 value1, Vector2 value2)
...
+        /// </summary>
+        /// <param name="value1">The first vector.</param>
+        /// <param name="value2">The second vector.</param>
+        /// <returns>The value of the z-coordinate from the cross product.</returns>
+        /// <remarks>
+        /// Return z-value = value1.X * value2.Y - value1.Y * value2.X
+        /// <see cref="Cross"/> is the same as taking the <see cref="Dot"/> with the second vector
+        /// that has been rotated 90-degrees.
+        /// </remarks>
+        [Intrinsic]
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector2.cs:341]: my understanding is that as long as you don't contribute a JIT change, then it won't get replaced by JIT with more optimal codegen

@tannergooding please confirm

```suggestion
```

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 2]
```diff
@@ -230,6 +230,45 @@ public void Vector2DotTest2()
             Assert.True(float.IsNegativeInfinity(actual), "Vector2f.Dot did not return the expected value.");
         }
 
+        // A test for Cross (Vector2f, Vector2f)
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:233]: I am aware of the fact that other tests in this class use similar comments, but I don't believe they add any value.

```suggestion
```

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 3]
```diff
@@ -230,6 +230,45 @@ public void Vector2DotTest2()
...
+        [Fact]
+        public void Vector2CrossTest()
+        {
+            Vector2 a = new Vector2(1.0f, 2.0f);
+            Vector2 b = new Vector2(-4.0f, 3.0f);
+
+            float expected = 11.0f;
+            float actual;
+
+            actual = Vector2.Cross(a, b);
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:243]:
```suggestion
            float actual = Vector2.Cross(a, b);
```

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 4]
```diff
@@ -230,6 +230,45 @@ public void Vector2DotTest2()
...
+            Assert.True(MathHelper.Equal(expected, actual), "Vector2f.Cross did not return the expected value.");
+        }
+
+        // A test for Cross (Vector2f, Vector2f)
+        // Cross test for parallel vector
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:248]:
```suggestion
```

File: src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs [hunk 5]
```diff
@@ -230,6 +230,45 @@ public void Vector2DotTest2()
...
+            Vector2 a = new Vector2(1.55f, 1.55f);
+            Vector2 b = new Vector2(-1.55f, -1.55f);
+
+            float expected = 0.0f;
+            float actual = Vector2.Cross(a, b);
+            Assert.Equal(expected, actual);
+        }
+
+        // A test for Cross (Vector2f, Vector2f)
+        // Cross test with specail float values
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector2Tests.cs:261]:
```suggestion
```

File: src/libraries/System.Numerics.Vectors/tests/Vector4Tests.cs [hunk 6]
```diff
@@ -244,6 +244,39 @@ public void Vector4DotTest1()
             Assert.True(MathHelper.Equal(0.0f, actual), "Vector4f.Dot did not return the expected value.");
         }
 
+        // A test for Cross (Vector4f, Vector4f)
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector4Tests.cs:247]:
```suggestion
```

File: src/libraries/System.Numerics.Vectors/tests/Vector4Tests.cs [hunk 7]
```diff
@@ -244,6 +244,39 @@ public void Vector4DotTest1()
...
+            Vector4 a4 = new Vector4(a3, 2.0f);
+            Vector4 b4 = new Vector4(b3, 3.0f);
+            Vector4 e4 = new Vector4(e3, a4.W * b4.W);
+
+            Vector4 actual = Vector4.Cross(a4, b4);
+            Assert.True(MathHelper.Equal(e4, actual), "Vector4f.Cross did not return the expected value.");
+        }
+
+        // A test for Cross (Vector4f, Vector4f)
+        // Cross test of the same vector
```
Comment (adamsitnik) [inline src/libraries/System.Numerics.Vectors/tests/Vector4Tests.cs:264]:
```suggestion
```

**Review by adamsitnik — CHANGES_REQUESTED** [review]
Overall it LGTM, but we need to apply some minor cleanup before merging.

@AlexRadch thank you for your contribution!

**Comment by PavielKraskouski** [comment]
I'm keeping this pull request active.

**Pull request closed without merging**